### PR TITLE
Understand & resolve type signatures

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
This PR makes the compiler look into the type signatures for the first time ever.
It also shows off the new system by making the suggestion builder use it instead of trying to badly (but also surprisingly well for the amount of info available to it!) get it itself.

### Important Notes
Some of this logic (i.e. the places where we bend over backwards to only expose types when explicitly asked for) will have to be removed when sum types get actual runtime teeth and we resolve the conflicts in stdlib. This PR is designed to not break any existing code.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ x If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala [hunk 1]
```diff
@@ -46,6 +46,10 @@ case class BindingsMap(
     */
   var exportedSymbols: Map[String, List[ResolvedName]] = Map()
 
+  /** Symbols exported by [[currentModule]].
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala]: The documentation comment is exactly the same as the comment at `exportedSymbols`. However I believe there is going to be some difference, right?

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala [hunk 2]
```diff
@@ -508,9 +511,31 @@ class SuggestionBuilderTest extends AnyWordSpecLike with Matchers {
...
-              name          = "MyAtom",
+              name          = "Variant_1",
               arguments     = Seq(),
-              returnType    = "Unnamed.Test.MyAtom",
+              returnType    = "Unnamed.Test.Variant_1",
+              documentation = None
+            ),
+            Vector()
+          ),
+          Tree.Node(
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala:521]: My understanding is that the `SuggestionBuilderTest` is being updated to new, better replies available thanks to _understanding & resolving type signatures_. Great!

File: engine/runtime/src/test/scala/org/enso/compiler/test/semantic/TypeSignaturesTest.scala [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+      langCtx.getCompiler.run(module)
+      module.getIr
+    }
+  }
+
+  private def getSignature(
+    module: IR.Module,
+    methodName: String
+  ): IR.Expression = {
+    val m = module.bindings.find {
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/scala/org/enso/compiler/test/semantic/TypeSignaturesTest.scala:156]: If we wanted to expose signatures to Enso programmers. For example via `Meta.signature`, then the implementation would be the same as code in this method, right?

File: engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala [hunk 4]
```diff
@@ -120,14 +127,27 @@ case class BindingsMap(
       if (newSymbols.exists { case (_, v) => v.isEmpty }) {
         None
       } else {
-        bindings.exportedSymbols = newSymbols.map { case (k, v) => k -> v.get }
+        bindings.allExportedSymbols = newSymbols.map { case (k, v) =>
+          k -> v.get
+        }
+        bindings.exportedSymbols = bindings.allExportedSymbols.flatMap {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala]: As far as I can tell `exportedSymbols` are always computed as `allExportedSymbols.flatMap`...

File: engine/runtime/src/main/scala/org/enso/compiler/phase/ExportsResolution.scala [hunk 5]
```diff
@@ -205,6 +213,14 @@ class ExportsResolution {
       ).flatten.groupBy(_._1).map { case (m, names) =>
         (m, names.flatMap(_._2).distinct)
       }
+
+      bindings.exportedSymbols = bindings.allExportedSymbols.flatMap {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/phase/ExportsResolution.scala]: This is a second time `exportedSymbols` are computed as `allExportedSymbols.flatMap`...

I am no Scala guru, but shouldn't this be achieved via Scala's _lazy value_? Something like:
```scala
class ExportedSymbols(allExportedSymbols) {
  lazy val exportedSymbols = allExportedSymbols.flatMap(...)
}
```

File: engine/runtime/src/main/scala/org/enso/compiler/context/SuggestionBuilder.scala [hunk 6]
```diff
@@ -375,57 +337,38 @@ final class SuggestionBuilder[A: IndexedSource](val source: A) {
...
-          args :+ TypeArg.Application(appFunction, appArgs.toVector)
-        case tname: IR.Name =>
-          val typeName = tname.name
-          val tdef = resolveTypeName(bindings, typeName)
-            .getOrElse(TypeArg.Value(QualifiedName.simpleName(typeName)))
-          args :+ tdef
-        case _ =>
-          args
-      }
+    def go(expr: IR.Expression): TypeArg = expr match {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/context/SuggestionBuilder.scala:364]: `SuggestionBuilder` no longer needs `BindingAnalysis` and the mapping from `IR.Expression` to `TypeArg` seems like straightforward 1:1 conversion. Great simplification!

**Review by JaroslavTulach — APPROVED** [review]
`SuggestionBuilder` seems simplified. Newly added tests are sign of improved behavior. Good!

File: engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala [hunk 7]
```diff
@@ -93,6 +93,8 @@ trait PackageRepository {
     */
   def registerModuleCreatedInRuntime(module: Module): Unit
 
+  def registerSyntheticPackage(namespace: String, name: String): Unit
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala:97]: Can you document this public method?

File: engine/runtime/src/main/scala/org/enso/compiler/context/SuggestionBuilder.scala [hunk 8]
```diff
@@ -668,7 +617,8 @@ object SuggestionBuilder {
       *
       * @param signature the list of types defining the function
       */
-    case class Function(signature: Vector[TypeArg]) extends TypeArg
+    case class Function(arguments: Vector[TypeArg], result: TypeArg)
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/compiler/context/SuggestionBuilder.scala:643]: missing docs update?

File: engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/TypeFunctions.scala [hunk 9]
```diff
@@ -150,42 +150,54 @@ case object TypeFunctions extends IRPass {
...
+    }
+  }
+
   /** Resolves a known typing function to its IR node.
     *
     * @param prefix the application to resolve
     * @return the IR node representing `prefix`
     */
-  def resolveKnownFunction(prefix: IR.Application.Prefix): IR.Expression = {
+  def resolveKnownFunction(
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/TypeFunctions.scala:167]: missing docs update

File: engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/TypeNames.scala [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+import org.enso.compiler.core.IR
+import org.enso.compiler.core.ir.MetadataStorage.ToPair
+import org.enso.compiler.data.BindingsMap
+import org.enso.compiler.data.BindingsMap.TypeResolution
+import org.enso.compiler.pass.IRPass
+import org.enso.compiler.pass.analyse.BindingAnalysis
+
+import scala.annotation.unused
+
+/** Resolves and desugars referent name occurences in type positions.
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/TypeNames.scala:13]: what is a "referent name"?

File: engine/runtime/src/main/scala/org/enso/compiler/phase/ExportsResolution.scala (hunk 5 above)
Comment (hubertp) [reply engine/runtime/src/main/scala/org/enso/compiler/phase/ExportsResolution.scala]: `allExportedSymbols` is a var, so no

**Review by hubertp — COMMENTED** [review]
I would probably add a test that ensures that a slightly non-trivial type like
```
foo: A | B | C -> D | E -> F
```
is inferred as desired.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/AtomConstructor.java [hunk 11]
```diff
@@ -43,6 +44,8 @@ public final class AtomConstructor implements TruffleObject {
   private @CompilerDirectives.CompilationFinal Atom cachedInstance;
   private @CompilerDirectives.CompilationFinal Function constructorFunction;
 
+  private final Type type;
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/AtomConstructor.java:45]: Is  there 1:1 mapping between `AtomConstructor` and `Type`, right now? E.g. it is more a refactoring and moving type-related code out of `AtomConstructor`.

How will the relation ship between constructor and type look like for `Vector<Boolean>` and `Vector<Text>`? Will the constructors be `==`? Now? In the future?

**Comment by JaroslavTulach** [comment]
> fix all of base in [ca352cd](https://github.com/enso-org/enso/pull/3511/commits/ca352cde8b080cc3b6d6126f20acfc3f5d649939)

Great!

**Pull request closed without merging**